guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+3. Define a `TimeoutError` struct and require that our generic error type can be
+   constructed from a `TimeoutError` using `TimeoutError: Into<Error>`.
+
+While option 3 might seem like it is the most flexible it isn't great since it
+requires users using a custom error type to manually implement
+`From<TimeoutError> for MyError`. That quickly becomes tedious when using lots
+of middleware that each have their own error type.
+
+Option 2 would mean defining an enum like this:
+
+```rust
+enum TimeoutError<Error> {
+    // Variant used if we hit the timeout
+    Timeout(InnerTimeoutError),
+    // Variant used if the inner service produced an error
+    Service(Error),
+}
+```
+
+While this seems ideal on the surface as we're not losing any type information
+and can use `match` to get at the exact error, the approach has three issues:
+
+1. In practice its common to nest lots of middleware. That would make the final
+   error enum very large. Its not unlikely to look something like
+   `BufferError<RateLimitError<TimeoutError<MyError>>>`. Pattern matching on
+   such a type (to for example determine if the error is retry-able) is very
+   tedious.
+2. If we change the order our middleware are applied in we also change the final
+   error type meaning we have to update our pattern matches.
+3. There is also the possibility of the final error type being very large and
+   taking up a significant amount of space on the stack.
+
+With this we're left with option 1 which is to convert the inner service error
+into a boxed trait object like `Box<dyn std::error::Error + Send + Sync>`. That
+means we can combine multiple errors type into one. That has the following
+advantages:
+
+1. Our error handling is less fragile since changing the order middleware are
+   applied in wont change the final error type.
+2. The error type now has a constant size regardless how many middleware we've
+   applied.
+3. Extracting the error no longer requires a big `match` but can instead be done
+   with `error.downcast_ref::<Timeout>()`.
+
+However it also has the following downsides:
+
+1. As we're using dynamic downcasting the compiler can no longer guarantee that
+   we're exhaustively checking for every possible error type.
+2. Creating an error now requires an allocation. In practice we expect errors to
+   be infrequent and therefore this shouldn't be a problem.
+
+Which option you prefer is a matter of personal preference. Both have their
+advantages and disadvantages. However the pattern that we've decided to use in
+Tower is boxed trait objects. You can find the original discussion
+[here](https://github.com/tower-rs/tower/issues/131).
+
+For our `Timeout` middleware that means we need to create a struct that
+implements `std::error::Error` such that we can convert it into a `Box<dyn
+std::error::Error + Send + Sync>`. We also have to require that the inner
+service's error type implements `Into<Box<dyn std::error::Error + Send +
```

Comment:
And one note here is that we must always use `Into` if some middleware use `From` and other use `Into` it will break the stack in horrible to understand ways.

Reply:
Good point! std mentions that. Gonna add a link to it.